presto-hive/src/main/java/com/facebook/presto/hive/rule/AbstractFilterPushdown.java

```diff
@@ -0,0 +1,720 @@
+import static com.google.common.collect.ImmutableBiMap.toImmutableBiMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.intersection;
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractFilterPushdown
+        implements ConnectorPlanOptimizer
+{
+    private static final ConnectorTableLayout EMPTY_TABLE_LAYOUT = new ConnectorTableLayout(
+            new ConnectorTableLayoutHandle() {},
+            Optional.empty(),
+            TupleDomain.none(),
+            Optional.empty(),
+            Optional.empty(),
+            Optional.empty(),
+            emptyList());
+
+    protected final RowExpressionService rowExpressionService;
+    protected final StandardFunctionResolution functionResolution;
+    protected final FunctionMetadataManager functionMetadataManager;
+
+    public AbstractFilterPushdown(
+            RowExpressionService rowExpressionService,
+            StandardFunctionResolution functionResolution,
+            FunctionMetadataManager functionMetadataManager)
+    {
+        this.rowExpressionService = requireNonNull(rowExpressionService, "rowExpressionService is null");
+        this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
+        this.functionMetadataManager = requireNonNull(functionMetadataManager, "functionMetadataManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return doOptimize(maxSubplan, session, idAllocator);
+    }
+
+    public ConnectorPushdownFilterResult pushdownFilter(
+            ConnectorSession session,
+            ConnectorMetadata metadata,
+            ConnectorTableHandle tableHandle,
+            RowExpression filter,
+            Optional<ConnectorTableLayoutHandle> currentLayoutHandle)
+    {
+        Result result = checkConstantBooleanExpression(rowExpressionService, functionResolution, filter, currentLayoutHandle, metadata, session);
+        if (result.getConnectorPushdownFilterResult() != null) {
+            return result.getConnectorPushdownFilterResult();
+        }
+
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+        TupleDomain<ColumnHandle> entireColumnDomain = getEntireColumnDomain(result, columnHandles);
+
+        if (currentLayoutHandle.isPresent()) {
+            entireColumnDomain = intersectionWithPartitionColumnPredicate(currentLayoutHandle.get(), entireColumnDomain);
+        }
+
+        Constraint<ColumnHandle> constraint = new Constraint<>(entireColumnDomain);
+
+        // Extract deterministic conjuncts that apply to partition columns and specify these as Constraint#predicate
+        constraint = extractDeterministicConjuncts(
```

Comment:
Can we make this function non static so that we dont' have to pass the 3 class member fields (rowExpressionService,
                functionResolution,
                functionMetadataManager,)